### PR TITLE
wip:add native build

### LINKED ISSUE #371: Is it possible to support Scala Native?
Is it possible to support Scala Native?

Fastparse supports scala native(2.12, 2.13) but does not support scala 3. On the other hand, catsparse supports scala 3, but does not support scala native😖

### PR DESCRIPTION
**Pull request by i10416**
close #371 

Waiting for non-snapshot cats-core_native0.4_3 published ~which depends on~
- ~https://github.com/typelevel/scalacheck/pull/868~
and test dependencies for native 3.x, which depends on
- ~https://github.com/typelevel/scalacheck/pull/868~
- ~https://github.com/scalameta/munit/pull/477~
~and they are waiting for the compiler flag to be introduced in Scala 3.1.2.~

Now, cats(snapshot version) supports Scala Native x Scala 3
- https://github.com/typelevel/cats/pull/4228

### CODE REVIEW

File: core/shared/src/test/scala/cats/parse/BitSetTest.scala [hunk 1]
```diff
@@ -24,11 +24,11 @@ package cats.parse
 import org.scalacheck.Prop.forAll
 
 class BitSetTest extends munit.ScalaCheckSuite {
-  test("isScalaJs/isScalaJvm is consistent") {
+  test("isScalaJs/isScalaJvm/isScalaNative is consistent") {
     // This will need to be updated if we ever add scala-native
```
Comment (johnynek) [inline core/shared/src/test/scala/cats/parse/BitSetTest.scala:28]: can remove the comment.

**Comment by johnynek** [comment]
looks like you need to run `prePR` or `fmt` or something.

**Comment by i10416** [comment]
follow up #384

**Comment by i10416** [comment]
Hmm, cats-core_native2.11 is not published...🤔

**Comment by johnynek** [comment]
can you disable the native build for 2.11?

**Comment by i10416** [comment]
Yes, of course :)

**Comment by i10416** [comment]
https://github.com/sbt/sbt/releases/tag/v1.7.0-M1

**Comment by armanbilge** [comment]
@i10416 thanks for the heads-up! If you're interested, the plan is to integrate that with sbt-typelevel in https://github.com/typelevel/sbt-typelevel/issues/138.

**Comment by johnynek** [comment]
@i10416 I'd like to publish the next version soon. Do you want to try to get the native build ready to be in that release?

**Comment by i10416** [comment]
@johnynek I am not in a hurry, so feel free to publish the next release😉 I think(or hope) Scala 3.1.2 will be available until the one after the next cats-parse release and I will finish this PR until then.

**Comment by i10416** [comment]
@armanbilge Thanks! I will take a look at the linked issue.

**Comment by armanbilge** [comment]
Actually, I think we should forget about Scala 3.0 and update everything everyone to Scala 3.1. Lots of folks seem to agree, including the munit maintainers. Then we don't need to wait for Scala 3.1.2. See:
- https://github.com/typelevel/scalacheck/pull/879
- https://github.com/scalameta/munit/pull/477#issuecomment-1068501049

**Comment by i10416** [comment]
@armanbilge 
Ah, I see. I think that's nice news for most of Scala 3 users, as it unlock downstream libraries. In addition, I guess those who have already  adapted Scala 3.x are likely to be Scala power user, willing to update to latest Scala 3.x version.

Btw, I'm a bit busy in April and progress may delay.

File: core/shared/src/test/scala/cats/parse/BitSetTest.scala [hunk 2]
```diff
@@ -26,10 +26,11 @@ import org.scalacheck.Prop.forAll
 class BitSetTest extends munit.ScalaCheckSuite {
   // TODO: Remove isScalaJs/isScalaJvm in next minor version update. See https://github.com/typelevel/cats-parse/issues/391.
   test("isScalaJs/isScalaJvm is consistent") {
-    // This will need to be updated if we ever add scala-native
-    assert(!(BitSetUtil.isScalaJs && BitSetUtil.isScalaJvm))
-    assert(BitSetUtil.isScalaJs || BitSetUtil.isScalaJvm)
-    assert(BitSetUtil.isScalaJs ^ BitSetUtil.isScalaJvm)
+    if (BitSetUtil.isScalaJs || BitSetUtil.isScalaJvm) {
+      assert(!(BitSetUtil.isScalaJs && BitSetUtil.isScalaJvm))
+      assert(BitSetUtil.isScalaJs || BitSetUtil.isScalaJvm)
```
[the remark this answers is not in the document]
Comment (i10416) [reply core/shared/src/test/scala/cats/parse/BitSetTest.scala]: done

File: core/shared/src/test/scala/cats/parse/ParserTest.scala [hunk 3]
```diff
@@ -610,7 +610,7 @@ class ParserTest extends munit.ScalaCheckSuite {
 
   import ParserGen.{arbParser0, arbParser, biasSmall, genString, arbString}
 
-  val tests: Int = if (BitSetUtil.isScalaJs) 50 else 2000
+  val tests: Int = if (BitSetUtil.isScalaJvm) 2000 else 50
```
Comment (armanbilge) [inline core/shared/src/test/scala/cats/parse/ParserTest.scala:613]: Can we keep this condition `isScalaJs`? Is Scala Native runtime performance so slow?

File: core/shared/src/test/scala/cats/parse/RadixNodeTest.scala [hunk 4]
```diff
@@ -25,7 +25,7 @@ import org.scalacheck.{Gen, Prop}
 import org.scalacheck.Prop.forAll
 
 class RadixNodeTest extends munit.ScalaCheckSuite {
-  val tests: Int = if (BitSetUtil.isScalaJs) 50 else 20000
+  val tests: Int = if (BitSetUtil.isScalaJvm) 20000 else 50
```
Comment (armanbilge) [inline core/shared/src/test/scala/cats/parse/RadixNodeTest.scala:28]: Same here.

File: build.sbt [hunk 5]
```diff
@@ -79,6 +86,22 @@ lazy val core = crossProject(JSPlatform, JVMPlatform)
     crossScalaVersions := (ThisBuild / crossScalaVersions).value.filterNot(_.startsWith("2.11")),
     coverageEnabled := false
   )
+  .nativeSettings(
+    resolvers ++= Seq(
+      "Sonatype OSS Snapshots" at "https://oss.sonatype.org/content/repositories/snapshots",
+      "Sonatype OSS Snapshots s01" at "https://s01.oss.sonatype.org/content/repositories/snapshots"
+    ),
+    crossScalaVersions := (ThisBuild / crossScalaVersions).value.filterNot(_.startsWith("2.11")),
+    mimaPreviousArtifacts := Set.empty,
```
Comment (armanbilge) [inline build.sbt:95]:
```suggestion
    tlMimaVersionIntroduced := List("2.12", "2.13", "3").map(_ -> "0.3.8").toMap,
```

Comment (armanbilge) [reply build.sbt]: Sorry, I wrote it wrong! It's just `tlVersionIntroduced`, no mima 😅

File: core/shared/src/test/scala/cats/parse/ParserTest.scala (hunk 3 above)
Comment (i10416) [reply core/shared/src/test/scala/cats/parse/ParserTest.scala:613]: Actually, when I tested it before, I felt scala native is slower than JVM. I will check it again and if it is not so slower, I will put those conditions back.

Comment (armanbilge) [reply core/shared/src/test/scala/cats/parse/ParserTest.scala:613]: No problem. Well that's disappointing 😂

File: core/shared/src/test/scala/cats/parse/RadixNodeTest.scala (hunk 4 above)
Comment (i10416) [reply core/shared/src/test/scala/cats/parse/RadixNodeTest.scala:28]: It takes about 5 times longer to run this test(RadixNodeTest) on native than on JVM. Is this acceptable speed?

File: core/shared/src/test/scala/cats/parse/ParserTest.scala (hunk 3 above)
Comment (i10416) [reply core/shared/src/test/scala/cats/parse/ParserTest.scala:613]: Yeah, perhaps there are some performance bottle necks in Scala Native runtime.

Comment (armanbilge) [reply core/shared/src/test/scala/cats/parse/ParserTest.scala:613]: Ok, thanks for trying! Let's just leave it.

Anyway, the Native tests are not running with full optimizations enabled, perhaps that will be better.

Comment (i10416) [reply core/shared/src/test/scala/cats/parse/ParserTest.scala:613]: Ah, that may be the case. I don't think we should run every ci tests with full optimizations, which take quite long.

Comment (i10416) [reply core/shared/src/test/scala/cats/parse/ParserTest.scala:613]: And it turned out that it takes long to run tests on Scala Native if we set tests 2000 :(

Comment (i10416) [reply core/shared/src/test/scala/cats/parse/ParserTest.scala:613]: I hope Scala Native gets faster in future release🙁

File: core/shared/src/test/scala/cats/parse/RadixNodeTest.scala (hunk 4 above)
Comment (regadas) [reply core/shared/src/test/scala/cats/parse/RadixNodeTest.scala:28]: Wow i wasn’t expecting such a difference. Out of scope and just curious … any clue to where this degradation is coming from?

Comment (armanbilge) [reply core/shared/src/test/scala/cats/parse/RadixNodeTest.scala:28]: Sadly I'm also hitting big performance deficits in https://github.com/typelevel/spire/pull/1180. In some cases Scala.js seems to be 20x faster than Scala Native.

File: build.sbt [hunk 6]
```diff
@@ -79,6 +86,22 @@ lazy val core = crossProject(JSPlatform, JVMPlatform)
...
+    ),
+    crossScalaVersions := (ThisBuild / crossScalaVersions).value.filterNot(_.startsWith("2.11")),
+    tlVersionIntroduced := List("2.12", "2.13", "3").map(_ -> "0.3.8").toMap,
+    libraryDependencies := {
+      if (scalaVersion.value.startsWith("3.1")) {
+        libraryDependencies.value.filterNot(_ == Dependencies.cats) ++ Seq(
+          Dependencies.catsSnapshot.value
+        )
+      } else { libraryDependencies.value }
+    },
```
Comment (armanbilge) [inline build.sbt:102]: https://github.com/typelevel/cats/releases/tag/v2.8.0

**Pull request closed without merging**